elide-core/src/main/java/com/yahoo/elide/core/pagination/Pagination.java

```diff
@@ -88,23 +76,65 @@ public static Pagination parseQueryParams(final MultivaluedMap<String, String> q
-        }
+        if (isPageBased(pageData)) {
+
+            // Page-based pagination strategy
+            int limit = pageData.get(PaginationKey.size);
+            if (limit > MAX_PAGE_LIMIT) {
+                throw new InvalidValueException("page[size] value must be less than or equal to " + MAX_PAGE_LIMIT);
+            } else if (limit < 0) {
+                throw new InvalidValueException("page[size] must contain a positive value.");
+            }
+
+            int pageNumber = pageData.containsKey(PaginationKey.number) ? pageData.get(PaginationKey.number) : 1;
+            if (pageNumber < 1) {
+                throw new InvalidValueException("page[number] must contain a positive value.");
+            }
+
+            int computedOffset = pageNumber > 0 ? (pageNumber - 1) * limit : 0;
+            return new Pagination(computedOffset, limit);
+
+        } else if (isOffsetBased(pageData)) {
+
+            // Offset-based pagination strategy
+            int limit = pageData.containsKey(PaginationKey.limit)
+                    ? pageData.get(PaginationKey.limit) : DEFAULT_PAGE_LIMIT;
+
+            int offset = pageData.containsKey(PaginationKey.offset) ? pageData.get(PaginationKey.offset) : 0;
+
+            if (limit < 0 || offset < 0) {
+                throw new InvalidValueException("page[offset] and page[limit] must contain positive values.");
+            }
+            return new Pagination(offset, limit);
 
-        if (isPageBased && !pageData.containsKey(PaginationKey.offset)) {
-            int page = pageData.get(PaginationKey.page);
-            pageData.put(PaginationKey.offset, (page > 0 ? page - 1 : 0) * limit);
         }
-        if (!pageData.containsKey(PaginationKey.offset)) {
-            pageData.put(PaginationKey.offset, 0);
+
+        // Default Pagination
+        return DEFAULT_PAGINATION;
+    }
+
+    private static boolean isPageBased(Map<PaginationKey, Integer> pageData) {
+        if (pageData.containsKey(PaginationKey.size)) {
+            if (pageData.size() == 1 || pageData.containsKey(PaginationKey.number)) {
+                // Either Only contains page[size] in params
+                // Or contains Both page[size] and page[number]
+                return true;
+            } else {
+                throw new InvalidValueException("Invalid usage of pagination parameters.");
+            }
+        } else if (pageData.containsKey(PaginationKey.number)) {
+            // Contains page[number] but not page[size]
+            throw new InvalidValueException("page[size] parameter is required for page-based pagination.");
         }
-        // offset, page, limit
-        return new Pagination(pageData.get(PaginationKey.offset), limit);
+        return false;
+    }
+
+    private static boolean isOffsetBased(Map<PaginationKey, Integer> pageData) {
+        return  (pageData.containsKey(PaginationKey.limit) || pageData.containsKey(PaginationKey.offset));
```

Comment:
If that was the case, it would have thrown an exception in the call to isPageBased function